### PR TITLE
[ci][test-data/1] move upload_build_info into rayci

### PR DESCRIPTION
**Pull request by can-anyscale**
Move the `upload_build_info` login into rayci itself. We do this by creating a shared mount to share the bazel event logs across containers, then run the upload_build_info on the rayci container (forge).

Test:
- CI
- postmerge: https://buildkite.com/ray-project/postmerge/builds/2406

### CODE REVIEW

File: ci/build/upload_build_info.sh [hunk 1]
```diff
@@ -3,11 +3,14 @@
 # Cause the script to exit if a single command fails.
 set -ex
 
+chmod -R a+r /tmp/bazel_event_logs
```
Comment (can-anyscale) [inline ci/build/upload_build_info.sh:6]: this is needed now that the mount dir is shared across multiple dockers

File: .buildkite/others.rayci.yml [hunk 2]
```diff
@@ -17,12 +17,14 @@ steps:
         --only-tags doctest
         --except-tags gpu
         --parallelism-per-worker 3
+        --upload
```
Comment (aslonnie) [inline .buildkite/others.rayci.yml:20]: why is this required? why only for doctest?

File: ci/build/upload_build_info.sh (hunk 1 above)
Comment (aslonnie) [reply ci/build/upload_build_info.sh:6]: why? how does multi-container mounting has anything to do with mod bits?

File: ci/build/upload_build_info.sh [hunk 3]
```diff
@@ -3,11 +3,14 @@
...
+
 readonly PIPELINE_POSTMERGE="0189e759-8c96-4302-b6b5-b4274406bf89"
 readonly PIPELINE_CIV1_BRANCH="0183465b-c6fb-479b-8577-4cfd743b545d"
 if [[
   "${BUILDKITE_PIPELINE_ID:-}" != "${PIPELINE_POSTMERGE}" && 
-  "${BUILDKITE_PIPELINE_ID:-}" != "${PIPELINE_CIV1_BRANCH}"
+  "${BUILDKITE_PIPELINE_ID:-}" != "${PIPELINE_CIV1_BRANCH}" &&
+  "${BUILDKITE_BRANCH:-}" != "master"
```
Comment (aslonnie) [inline ci/build/upload_build_info.sh:13]: why skip non-master branches?

File: ci/build/upload_build_info.sh [hunk 4]
```diff
@@ -19,8 +22,10 @@ RAY_DIR=$(cd "${ROOT_DIR}/../../"; pwd)
 cd "${RAY_DIR}"
 
 cleanup() {
-  # Cleanup the directory because macOS file system is shared between builds.
-  rm -rf /tmp/bazel_event_logs
+  if [[ "$OSTYPE" == "darwin"* ]]; then
+    # Cleanup the directory because macOS file system is shared between builds.
+    rm -rf /tmp/bazel_event_logs
+  fi
```
Comment (aslonnie) [inline ci/build/upload_build_info.sh:28]: should we just do the remove on all systems? why change the behavior?

**Review by aslonnie — COMMENTED** [review]
I am overall not a fan of this change.

this is crossing bazel / container boundary, using bazel log files as protocol, but for your own convenience to implement this feature.

I think a better implementation would be:

- upload the logs to s3
- notify the service that some test is done
- and the service parse the logs offline asyncly

File: ci/build/upload_build_info.sh (hunk 1 above)
Comment (can-anyscale) [reply ci/build/upload_build_info.sh:6]: ray_ci is run under forge which doesn't have sudo permissions; while the container used for running tests is run as sudo

File: .buildkite/others.rayci.yml (hunk 2 above)
Comment (can-anyscale) [reply .buildkite/others.rayci.yml:20]: it's for controlling the rollout of this feature instead of rolling out for every test jobs

File: ci/build/upload_build_info.sh (hunk 4 above)
Comment (can-anyscale) [reply ci/build/upload_build_info.sh:28]: also not changing the behavior, the existing implementation supposes to delete it only on mac

**Comment by can-anyscale** [comment]
Let's discuss the design in person maybe tomorrow.

File: ci/build/upload_build_info.sh (hunk 4 above)
Comment (aslonnie) [reply ci/build/upload_build_info.sh:28]: ah, ok. you can split this part into a separate small PR if you want to get this particular change in.

Comment (can-anyscale) [reply ci/build/upload_build_info.sh:28]: totally

File: ci/build/upload_build_info.sh [hunk 5]
```diff
@@ -13,20 +15,20 @@ if [[
...
 cd "${RAY_DIR}"
 
-mkdir -p /tmp/bazel_event_logs
+mkdir -p "$bazel_log_dir"
 
-./ci/build/get_build_info.py > /tmp/bazel_event_logs/metadata.json
+./ci/build/get_build_info.py > "$bazel_log_dir"/metadata.json
 
 # Keep cryptography/openssl in sync with `requirements/test-requirements.txt`
 pip install -q -c "${RAY_DIR}/python/requirements.txt" docker aws_requests_auth boto3 cryptography==38.0.1 PyOpenSSL==23.0.0
```
Comment (can-anyscale) [inline ci/build/upload_build_info.sh:33]: boto3 here requires an upgrade of awscli in rayci

File: ci/build/upload_build_info.sh [hunk 6]
```diff
@@ -3,6 +3,8 @@
 # Cause the script to exit if a single command fails.
 set -ex
 
+bazel_log_dir=${1:-"/tmp/bazel_event_logs"}
```
Comment (aslonnie) [inline ci/build/upload_build_info.sh:6]: could you change this to all caps please: `BAZEL_LOG_DIR`

File: ci/ray_ci/test_linux_tester_container.py [hunk 7]
```diff
@@ -37,7 +38,8 @@ def test_enough_gpus() -> None:
         assert False, "Should not raise an AssertionError"
 
 
-def test_run_tests_in_docker() -> None:
+@mock.patch.object(tempfile, "mkdtemp")
```
Comment (aslonnie) [inline ci/ray_ci/test_linux_tester_container.py:41]: why do you need to mock `mkdtemp`?

could you just inject the `get_artifact_mount()` in the test setup? then it can create a temp dir for real?

generally speaking, mocking an external rpc call (e.g. s3 api call) is okay, but we prefer to run filesystem calls for real in tests.

File: ci/ray_ci/tester_container.py [hunk 8]
```diff
@@ -41,6 +45,19 @@ def __init__(
...
+        """
+        Create a temporary directory in the current container to store bazel event logs
+        produced by the test runs. We do this by using the artifact mount directory from
+        the host machine as a shared directory between all containers.
+        """
+        artifact_host, artifact_container = self.get_artifact_mount()
+        bazel_log_dir_container = tempfile.mkdtemp(dir=artifact_container)
+        bazel_log_dir_host = bazel_log_dir_container.replace(
+            artifact_container, artifact_host
+        )
```
Comment (aslonnie) [inline ci/ray_ci/tester_container.py:58]: could you do a safer prefix-based replace?

if you make up a random dir name on the previous line, you do not need this replace anymore.

File: ci/ray_ci/tester_container.py [hunk 9]
```diff
@@ -41,6 +45,19 @@ def __init__(
...
             self.install_ray(build_type)
 
+    def _get_bazel_log_mount_dir(self) -> Tuple[str, str]:
+        """
+        Create a temporary directory in the current container to store bazel event logs
+        produced by the test runs. We do this by using the artifact mount directory from
+        the host machine as a shared directory between all containers.
+        """
+        artifact_host, artifact_container = self.get_artifact_mount()
+        bazel_log_dir_container = tempfile.mkdtemp(dir=artifact_container)
```
Comment (aslonnie) [inline ci/ray_ci/tester_container.py:55]: `mkdtemp` requires manual cleanup anyways.

you can just give it a proper random dir name? seems that there is no need to use `tempfile` package?

File: ci/ray_ci/tester_container.py [hunk 10]
```diff
@@ -65,23 +82,45 @@ def run_tests(
...
+        self.persist_test_results(bazel_log_dir_container)
+
         return all(exit == 0 for exit in exits)
 
+    def persist_test_results(self, bazel_log_dir: str) -> None:
+        logger.info("Uploading test results")
+        self._upload_build_info(bazel_log_dir)
+
+        # clean up so we don't pollute the host machine
+        shutil.rmtree(bazel_log_dir)
```
Comment (aslonnie) [inline ci/ray_ci/tester_container.py:102]: can we move this to a separate cleanup step instead?

File: ci/ray_ci/tester_container.py (hunk 9 above)
Comment (aslonnie) [reply ci/ray_ci/tester_container.py:55]: I feel that you can just make a random id for every test on init, and use it throughout this class instance, and then you do not really need this method anymore.

**Comment by can-anyscale** [comment]
Generate a random string to use as tmpdir. I do this in `run_tests` instead of init because an object can call `run_tests` multiple time and the tmpdir need to be different across test runs.

File: ci/ray_ci/tester_container.py [hunk 11]
```diff
@@ -41,6 +47,19 @@ def __init__(
...
+        """
+        Create a temporary directory in the current container to store bazel event logs
+        produced by the test runs. We do this by using the artifact mount directory from
+        the host machine as a shared directory between all containers.
+        """
+        tmp_dir = "".join(random.choice(string.ascii_lowercase) for _ in range(5))
+        artifact_host, artifact_container = self.get_artifact_mount()
+        bazel_log_dir_host = os.path.join(artifact_host, tmp_dir)
+        bazel_log_dir_container = os.path.join(artifact_container, tmp_dir)
+        os.mkdir(bazel_log_dir_container)
```
Comment (can-anyscale) [inline ci/ray_ci/tester_container.py:60]: this feel very similar with doing `tempfile.TemporaryDirectory` though, should I use that instead and piggy back on python to clean up the directory for me?

Comment (aslonnie) [reply ci/ray_ci/tester_container.py:60]: up to you. I think either is fine. using `tempfile.TemporaryDirectory` does not give us a lot of benefit here.

I personally feel that not using `tempfile` is slightly more reasonable. `tempfile.TemporaryDirectory` is supposed to be "private", meaning it is set up in a (secure) way that only the owner can access it, and it should not be mounted from one namespace/container into another.

File: ci/ray_ci/tester_container.py [hunk 12]
```diff
@@ -65,23 +84,46 @@ def run_tests(
 
         # divide gpus evenly among chunks
         gpu_ids = chunk_into_n(list(range(self.gpus)), len(chunks))
+        (bazel_log_dir_host, bazel_log_dir_container) = self._create_bazel_log_mount()
```
Comment (aslonnie) [inline ci/ray_ci/tester_container.py:87]: nit: you can drop the parens on unpacking.

File: ci/ray_ci/test_linux_tester_container.py [hunk 13]
```diff
@@ -153,5 +165,20 @@ def _mock_shard_tests(tests: List[str], workers: int, worker_id: int) -> List[st
         assert not container.run_tests(["bad_test"], [])
 
 
+def test_create_bazel_log_mount() -> None:
+    with tempfile.TemporaryDirectory() as tmpdir, mock.patch(
+        "random.choice",
+        return_value="a",
```
Comment (aslonnie) [inline ci/ray_ci/test_linux_tester_container.py:171]: maybe just allow sending in an id explicitly? and when it is None/empty, generate a random one?

File: ci/ray_ci/test_linux_tester_container.py [hunk 14]
```diff
@@ -134,7 +140,13 @@ def _mock_run_tests_in_docker(
     def _mock_shard_tests(tests: List[str], workers: int, worker_id: int) -> List[str]:
         return chunk_into_n(tests, workers)[worker_id]
 
-    with mock.patch(
+    with tempfile.TemporaryDirectory() as tmpdir, mock.patch(
+        "ci.ray_ci.linux_tester_container.LinuxTesterContainer.get_artifact_mount",
+        return_value=("/tmp/artifacts", tmpdir),
```
Comment (aslonnie) [inline ci/ray_ci/test_linux_tester_container.py:145]: why does it need this mocking? now you can send in a real temp dir in testing I guess?

**Review by aslonnie — APPROVED** [review]
